lib/search/fragmentation_scheme.ex

```diff
@@ -0,0 +1,152 @@
+defmodule Search.FragmentationScheme do
+  @doc """
+  Splits a binary into multiple binaries that satisfy limitations specified by opts.
+
+  If possible, splits the text on whitespace to preserve words. If that is impossible, splits text in between graphemes.
+
+  Supported options:
+
+  * `:max_size` - maximum byte_size of the output binaries. The output binaries may have size less or equal to that
+    value, which also should guarantee the sequence length after tokenization will be bounded by this value.
+  """
+  def split(text, opts \\ [])
+  def split("", _opts), do: []
+
+  def split(text, opts) when is_binary(text) do
+    case Keyword.get(opts, :max_size) do
+      nil ->
+        [text]
+
+      max_size ->
+        text
+        |> compute_splits({0, 0}, 0, max_size, [])
+        |> split_binary(0, text, [])
+    end
+  end
+
+  def recombine(chunks), do: Enum.join(chunks)
+
+  defp split_binary([], _start_idx, _binary, acc), do: Enum.reverse(acc)
+
+  defp split_binary([split_size | splits_tail], start_idx, binary, acc) do
+    current_part = binary_slice(binary, start_idx, split_size)
+    split_binary(splits_tail, start_idx + split_size, binary, [current_part | acc])
+  end
```

Comment:
We can simplify this a bit with body-recursion:

```elixir
defp split_binary([], ""), do: []

defp split_binary([size | sizes], string) do
  <<chunk::binary-size(^size), rest::binary>> = string
  [chunk | split_binary(sizes, rest)]
end
```

A body-recursive function is fine here, because we are building a list result. In such setting, a tail-recursive function with list accumulator uses basically the same amount of memory. More details in [this paragraph](https://erlang.org/documentation/doc-10.1/doc/efficiency_guide/myths.html#myth--tail-recursive-functions-are-much-faster-----than-recursive-functions) :)